drivers/spi/spi_nrf52_master.c

```diff
@@ -0,0 +1,452 @@
+		return -ENOTSUP;
+	}
+
+	if (config->operation & SPI_LINES_QUAD) {
+		SYS_LOG_ERR("Quad line mode unsupported\n");
+		return -ENOTSUP;
+	}
+
+	if (config->frequency > cfg->max_freq) {
+		SYS_LOG_ERR("Unsupported frequency");
+		return -ENOTSUP;
+	}
+
+	switch (config->frequency) {
+	case 125000:
+		frequency = SPIM_FREQUENCY_FREQUENCY_K125;
+		break;
+	case 250000:
+		frequency = SPIM_FREQUENCY_FREQUENCY_K250;
+		break;
+	case 500000:
+		frequency = SPIM_FREQUENCY_FREQUENCY_K500;
+		break;
+	case 1000000:
+		frequency = SPIM_FREQUENCY_FREQUENCY_M1;
+		break;
+	case 2000000:
+		frequency = SPIM_FREQUENCY_FREQUENCY_M2;
+		break;
+	case 4000000:
+		frequency = SPIM_FREQUENCY_FREQUENCY_M4;
+		break;
+	case 8000000:
+		frequency = SPIM_FREQUENCY_FREQUENCY_M8;
+		break;
+	case 16000000:
+		frequency = SPIM_FREQUENCY_FREQUENCY_M16;
+		break;
+	case 32000000:
+		frequency = SPIM_FREQUENCY_FREQUENCY_M32;
+		break;
+	default:
+		SYS_LOG_ERR("Unsupported frequency");
+		return -ENOTSUP;
+	}
+
+	spim->ENABLE = SPIM_ENABLE_ENABLE_Disabled;
+
+	spim->INTENCLR = 0xffffffff;
+
+	spim->SHORTS = 0;
+
+	spim->ORC = cfg->orc;
+
+	spim->TXD.LIST = 0;
+	spim->RXD.LIST = 0;
+
+	spim->TXD.MAXCNT = 0;
+	spim->RXD.MAXCNT = 0;
+
+	spim->EVENTS_END = 0;
+	spim->EVENTS_ENDTX = 0;
+	spim->EVENTS_ENDRX = 0;
+	spim->EVENTS_STOPPED = 0;
+	spim->EVENTS_STARTED = 0;
+
+	spim->FREQUENCY = frequency;
+
+	spim->CONFIG =
+	    (config->operation & SPI_TRANSFER_LSB)
+		? (SPIM_CONFIG_ORDER_LsbFirst << SPIM_CONFIG_ORDER_Pos)
+		: (SPIM_CONFIG_ORDER_MsbFirst << SPIM_CONFIG_ORDER_Pos);
+	spim->CONFIG |=
+	    (config->operation & SPI_MODE_CPOL)
+		? (SPIM_CONFIG_CPOL_ActiveLow << SPIM_CONFIG_CPOL_Pos)
+		: (SPIM_CONFIG_CPOL_ActiveHigh << SPIM_CONFIG_CPOL_Pos);
+	spim->CONFIG |=
+	    (config->operation & SPI_MODE_CPHA)
+		? (SPIM_CONFIG_CPHA_Trailing << SPIM_CONFIG_CPHA_Pos)
+		: (SPIM_CONFIG_CPHA_Leading << SPIM_CONFIG_CPHA_Pos);
+
+	if ((config->cs != NULL) || (cfg->psel.cs == CS_UNUSED)) {
+		/* Disable HW based CS */
+		spim->PSEL.CSN |= SPIM_PSEL_CSN_CONNECT_Msk;
+	} else {
+		/* Enable HW based CS */
+		spim->PSEL.CSN &= ~SPIM_PSEL_CSN_CONNECT_Msk;
+	}
+
+	spim->INTENSET = SPIM_INTENSET_END_Msk;
+
+	return 0;
+}
+
+static int transceive_current_buffer(volatile NRF_SPIM_Type *spim,
+				     struct spi_context *ctx)
+{
+	if (ctx->current_tx->len) {
+		__ASSERT_NO_MSG(ctx->current_tx->buf);
+		spim->TXD.MAXCNT = ctx->current_tx->len;
+		spim->TXD.PTR = (u32_t)ctx->current_tx->buf;
+	} else {
+		spim->TXD.MAXCNT = 0;
+	}
+
+	if (ctx->current_rx->len) {
+		__ASSERT_NO_MSG(ctx->current_rx->buf);
+		spim->RXD.MAXCNT = ctx->current_rx->len;
+		spim->RXD.PTR = (u32_t)ctx->current_rx->buf;
+	} else {
+		spim->RXD.MAXCNT = 0;
+	}
+
+	spim->TASKS_START = 1;
+	return 0;
+}
+
+static int transceive(struct spi_config *config, const struct spi_buf *tx_bufs,
+		      u32_t tx_count, struct spi_buf *rx_bufs, u32_t rx_count,
+		      bool asynchronous, struct k_poll_signal *signal)
+{
+	int ret;
```

Comment:
usually, pre-assigned variable declaration come first, so here ret would be the last